src/app-layer.c

```diff
@@ -648,6 +678,80 @@ void AppLayerRegisterGlobalCounters(void)
     StatsRegisterGlobalCounter("http.memcap", HTPMemcapGlobalCounter);
 }
 
+void AppLayerSetupCounters(uint8_t ipproto)
+{
+    AppProto alproto;
+    AppProto alprotos[ALPROTO_MAX];
+    uint8_t other_ipproto = (ipproto == IPPROTO_TCP) ? IPPROTO_UDP : IPPROTO_TCP;
+    const char *ipproto_suffix = (ipproto == IPPROTO_TCP) ? "_tcp" : "_udp";
+
+    AppLayerProtoDetectSupportedAppProtocols(alprotos);
+
+    for (alproto = 0; alproto < ALPROTO_MAX; alproto++) {
+        if (alprotos[alproto] == 1) {
+            char *str = "app_layer.flow.";
+            char *alproto_str = AppLayerGetProtoName(alproto);
+            int alproto_len = strlen(alproto_str) + 1;
+            uint8_t ipproto_map = FlowGetProtoMapping(ipproto);
+
+            if (AppLayerParserProtoIsRegistered(ipproto, alproto) &&
+                AppLayerParserProtoIsRegistered(other_ipproto, alproto))
+            {
+                applayer_counters[ipproto_map][alproto].name =
+                    SCMalloc(strlen(str) + alproto_len + strlen(ipproto_suffix));
+                if (applayer_counters[ipproto_map][alproto].name == NULL) {
+                    return;
+                }
+
+                snprintf(applayer_counters[ipproto_map][alproto].name,
+                         strlen(str) + alproto_len + strlen(ipproto_suffix),
+                         "%s%s%s", str, alproto_str, ipproto_suffix);
+            } else {
+                applayer_counters[ipproto_map][alproto].name =
+                    SCMalloc(strlen(str) + alproto_len);
+                if (applayer_counters[ipproto_map][alproto].name == NULL) {
+                    return;
+                }
+                snprintf(applayer_counters[ipproto_map][alproto].name,
+                         strlen(str) + alproto_len,
+                         "%s%s", str, alproto_str);
+            }
+        }
+    }
+}
+
+void AppLayerRegisterCounters(ThreadVars *tv, uint8_t ipproto)
+{
+    AppProto alproto;
+    AppProto alprotos[ALPROTO_MAX];
+
+    AppLayerProtoDetectSupportedAppProtocols(alprotos);
+
+    for (alproto = 0; alproto < ALPROTO_MAX; alproto++) {
+        if (alprotos[alproto] == 1) {
+            uint8_t ipproto_map = FlowGetProtoMapping(ipproto);
+            applayer_counters[ipproto_map][alproto].counter_id =
+                StatsRegisterCounter(applayer_counters[ipproto_map][alproto].name, tv);
+        }
+    }
+}
+
+void AppLayerDeRegisterCounters(uint8_t ipproto)
```

Comment:
This is not per thread, but it does have a name 'opposite' of a thread func: AppLayerRegisterCounters (per thread) vs AppLayerDeRegisterCounters (global)